### ISSUE TITLE
Tests using ScheduledProcess.shouldExit() are flaky

### ISSUE THREAD

**Issue by DartBot** [dart-lang/scheduled_test#8]
<a href="https://github.com/whesse"><img src="https://avatars.githubusercontent.com/u/4905639?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [whesse](https://github.com/whesse)**
_Originally opened as dart-lang/sdk#15654_

---

There is a flaky failure on MacOS of pkg/scheduled_test/test/scheduled_process, on subtest
&nbsp;&nbsp;&quot;remainingStdout returns the remaining stdout
&nbsp;&nbsp;&nbsp;after the lines consumed by nextLine&quot;

This is due to the fact that process.shouldExit(0) appears in the schedule after other scheduled items.  In fact, unless the process is known to wait on input before exiting, the use of process.shouldExit() is probably wrong.

The use of pumpEventQueue may make it correct in the case where everything is on the event queue, but IO is not all in the same thread.  The read events from process.stdout may not happen until after the process exit event happens, and the Dart thread may be allowed to run through the microtask event loop 20 times before the IO thread is scheduled again, and the read event fires.  In fact, the process exit is detected by a third thread, separate from the IO thread, so there is no guarantee about when the stdout can be read.

So even when the process is known to wait on some input, process.shouldExit has to be scheduled before that input is sent.
But then the schedule waits on the exit code before sending that input.
So we hang.

All we ever know when a process ends, after sending output, is that
1) At some point in the future, we will receive a exit event.
2) At some point in the future, we will receive a read event on stdout.
Nothing about the sequence, or that they will happen in the current microtask loop schedule, can be assumed.

I will note that the last paragraph should be put into the documentation of dart:io Process.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

---

Oof, do you have any suggestions for how we should proceed here?

Obviously, we need the ability to write reliable, deterministic tests that spawn processes and validate what they do. What's the best way to implement those tests?

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

---

A few more questions:

\* Why haven't we ever seen this when running local tests?

\* When you say "the process is known to wait on some input", are you referring to input being sent to stdin, or output emitted by stdout?

&nbsp;&nbsp;* If the former, how does this relate to the test in question, which doesn't interact with stdin at all?

&nbsp;&nbsp;* If the latter, in what sense is the process waiting?

\* If ScheduledProcess waits for stdin/stdout to close before it starts complaining about unexpected errors, would that fix the issue?

**Comment by DartBot** [comment]
<a href="https://github.com/skabet"><img src="https://avatars.githubusercontent.com/u/22043?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [skabet](https://github.com/skabet)**

---

This is the(a) test in question:

&nbsp;&nbsp;expectTestsPass("nextErrLine returns the next line of stderr from the "
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"process", () {
&nbsp;&nbsp;&nbsp;&nbsp;test('test', () {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;var process = startDartProcess(r'''
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;stderr.write("hello\n\nworld\n");
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;stderr.write("hi");
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;''');
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;expect(process.nextErrLine(), completion(equals('hello')));
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;expect(process.nextErrLine(), completion(equals('')));
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;expect(process.nextErrLine(), completion(equals('world')));
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;expect(process.nextErrLine(), completion(equals('hi')));
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;process.shouldExit(0);
&nbsp;&nbsp;&nbsp;&nbsp;});
&nbsp;&nbsp;});

Here shouldExit is called after the process is spawned and expect is issued for several async operations. Looking at shouldExit:

&nbsp;&nbsp;void shouldExit([int expectedExitCode]) {
&nbsp;&nbsp;&nbsp;&nbsp;if (_endScheduled) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;throw new StateError("shouldExit() or kill() already called.");
&nbsp;&nbsp;&nbsp;&nbsp;}

&nbsp;&nbsp;&nbsp;&nbsp;_endScheduled = true;
&nbsp;&nbsp;&nbsp;&nbsp;_taskBeforeEnd = currentSchedule.tasks.contents.last;
&nbsp;&nbsp;&nbsp;&nbsp;schedule(() {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_endExpected = true;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return _exitCode.then((exitCode) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if (expectedExitCode != null) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;expect(exitCode, equals(expectedExitCode));
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;});
&nbsp;&nbsp;&nbsp;&nbsp;}, "waiting for process '$description' to exit");
&nbsp;&nbsp;}

We can see that _endExpected is not set until this 'schedule'-callback is invoked, that is, some time in the future, after the current Dart message has been handled.

Now, in _handleExit we have the following 'guard' for when a process has terminated.

&nbsp;&nbsp;&nbsp;&nbsp;_process.then((p) => Chain.track(p.exitCode)).then((exitCode) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pumpEventQueue().then(...
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...
&nbsp;&nbsp;&nbsp;&nbsp;});

As the process can terminate at any point after it was started, it's possible it exited _before_ the 'schedule(...)' callback in shouldExit is executed. That means that, even though the event-queue is tried to be emptied, the _endExpected may be false and produce the following error:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if (!_endExpected) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;fail("Process '$description' ended earlier than scheduled "
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"with exit code $exitCode.");
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}

Even though the process terminated as expected.

To answer your questions:

> Oof, do you have any suggestions for how we should proceed here?

First off, I've a hard time understanding how scheduled_test is helping out here. It's making the execution order very implicit (though errors slightly more explicit). When I see code using 'scheduled_test', I have no easy way to follow the execution pattern as things are not called as part of the completion of a future, but rather when it's their time in the queue.

That being said, in this particular case, _endExpected should probably be set to true as soon as shouldExit is called. It should _never_ be the case that you'd should need to empty the event-loop (this can never be guaranteed and is bound to lead to flakyness).

IMO, to get reliable processes running, use their futures and streams (which you are already do), and don't relay on 'events' being done in the event-loop. Use futures for this.

> Why haven't we ever seen this when running local tests?

We have. Once in a 100-1000 runs. Particular scheduled_process_test has been flaky over time, along with some pub tests complaining about exit code.

As this is all part of a possible, but yet likely, scenario, don't expect it to just happen. Sometimes it's only easy to trigger if many tests are run at once (I normally use a 32 simultaneous runs in a loop, to trigger stuff like this. In this case, it appears to be easier to reproduce on Mac.

Also, remove all calls to pumpEventQueue should help (in general, it should never be required to perform such a task).

> When you say "the process is known to wait on some input", are you referring to input being sent to stdin, or output emitted by stdout?

This test is exiting as soon as it has printed. One could argue that it's not exiting before the print is done, thus before the stdout of the parent process was read. However, as the OS can buffer data, this is not the case. This particular process will exit at will. However, if the process was blocking until, let's say the byte '5' was received on stdin, it would be possible to block the exit of the child process, until 'shouldExit-schedule' has been invoked. However, this is not a fix, just a workaround in this particular case. It's more to point out that processes handling some http-servers, etc, may not be hit by this.

> If ScheduledProcess waits for stdin/stdout to close before it starts complaining about unexpected errors, would that fix the issue?

Yes and no. It may be, that reading stdout/stderr may delay the order of events enough for it to be harder to hit, but it won't fix the underlying issue that a process can exit before shouldExit is fully set up.

Let me know if you have more questions. I'd be glad to help get this sorted out!

**Comment by DartBot** [comment]
<a href="https://github.com/whesse"><img src="https://avatars.githubusercontent.com/u/4905639?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [whesse](https://github.com/whesse)**

---

It looks to me like there are two separate things you might want to test, at different points in the schedule.  Separating them makes everything much clearer:

1) The process does not exit until this code has executed.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This will not always be relevant - in the example test, the process can exit any time after it is created.  Only if the process waits for input before exiting would this test be desired.

2) The test can block, waiting for the process to exit, at this point, and will not be deadlocked.

Separating these two would make things simpler.  The first is just a scheduled step checking "The process is not exited", and the second is just a scheduled step that waits on the exitCode or the exit.
&nbsp;

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

---

> I've a hard time understanding how scheduled_test is helping out here.

This test in particular is testing a component of scheduled_test. In general, doing complicated assertions on processes' stdout/stderr in conjunction with providing input, serving HTTP requests, and checking exit codes is extremely difficult when using raw streams and futures. scheduled_test and scheduled_process make these tests much easier to read, write, and debug.

If you're interested in a deeper understanding of why scheduled_test is useful, it might be an informative exercise to rewrite a complicated pub test (e.g. a test in sdk/lib/_internal/pub/test/oauth2) to use raw futures and streams instead.

> remove all calls to pumpEventQueue should help (in general, it should never be required to perform such a task).

There are some cases where it's impossible to avoid using pumpEventQueue, by virtue of the fact that the code that wants to wait for asynchronous operations to complete may not have any sort of access to the asynchronous operations themselves. For example, in scheduled_test's MockClock implementation, the clock needs to wait for all asynchronous operations to complete between each tick, but it doesn't have access to any futures or streams that it can wait on.

That said, it's possible that many of the pumpEventQueue calls in the scheduled classes can hook into the current Schedule to get a more accurate representation of when there are no pending events. In addition, Zones now exist, which provide a better means of tracking unrelated callbacks. Once I get some time I'd like to refactor ScheduledTest to take advantage of those.

> Yes and no. It may be, that reading stdout/stderr may delay the order of events enough for it to be harder to hit, but it won't fix the underlying issue that a process can exit before shouldExit is fully set up.

My proposal is that _endExpected only be checked once both stdout and stderr have closed. So even if the process exits before shouldExit's scheduled callback runs, an error won't be raised unless stdout and stderr close before producing the expected output.

> It looks to me like there are two separate things you might want to test, at different points in the schedule.

You're correct about the assertions being made, but I don't think it provides value to test authors to separate them explicitly. Handling the "won't exit early" assertion automatically both removes the amount of boilerplate that goes into handling passing input to the process and allows the output in the case of a test failure to be clearer.

---

_Added [Accepted](../labels/Accepted) label._

**Comment by DartBot** [comment]
<a href="https://github.com/whesse"><img src="https://avatars.githubusercontent.com/u/4905639?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [whesse](https://github.com/whesse)**

---

I would like more detail on the last three paragraphs of the last comment.  If _endExpected is only checked after stdout and stderr have closed, then the only time that _endExpected could detect an error is if we have shouldExit after all the expected reads.  But if we put it immediately after the reads, and use pushevents to make _endExpected true, then the check is vacuous again.  I just find it hard to find a scenario where _endExpected detects an error that isn't detected by something else. 

&nbsp;It seems the whole check becomes sort of pointless then - it checks that process exit isn't signalled before shouldExit is called but after the streams close, but that error will also be detected by the streams not getting what they expect, since shouldExit will always be placed before the last stream read.  Unless we are going to place shouldExit after that, and rely on pumping to set _endExpected.

I really don't see why these two expectations should be happening at the same time in your tests, and be a single call: (A) We won't get an exit signal before this point, and (B) We can (and will) wait for the exit signal at this point.  

By making all this special handling of the exit timing checking, we are trying to artificially set the "no exit before this point" time (check A)  to be correctly virtually placed in the schedule at some earlier point, based on "shouldExit" being placed in the schedule at the time we want to wait on the exit signal (check B).

So please expand on the last three paragraphs, delaying the check of _endExpected and why the check for early exit is ever useful, and not redundant given the rest of the schedule.

I guess I am proposing that shouldExit should just be a wait on process exit, and that there will rarely be a need for a check for early exit, given that other checks in the schedule will fail.  And that determining the point in the schedule at which an exit is really "early", based on the point where we wait for exit in the schedule, is a complex and perhaps useless task.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

---

You're right; I hadn't thought it through thoroughly. It's probably sufficient just to make sure the debug output has an indication of when the subprocess exited.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

---

Fixed in dart-lang/sdk@4121e1df5362aaa163acbf7dcb1efb58c68be240.

---

_Added [Fixed](../labels/Fixed) label._

**Comment by DartBot** [comment]
<a href="https://github.com/anders-sandholm"><img src="https://avatars.githubusercontent.com/u/3276024?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [anders-sandholm](https://github.com/anders-sandholm)**

---

_Removed [Library-ScheduledTest](../labels/Library-ScheduledTest) label._
_Added [Pkg-ScheduledTest](../labels/Pkg-ScheduledTest) label._
